src/Accelerators/NNPA/Conversion/ONNXToZHigh/ONNXLegalityCheck.cpp

```diff
@@ -612,6 +614,11 @@ bool isSuitableForZDNN<ONNXGRUOp>(ONNXGRUOp op) {
   Value R = op.R();
   Value B = op.B();
 
+  // Check direction.
+  if ((direction != FORWARD) && (direction != REVERSE) &&
```

Comment:
A few general comments:
 
 ```auto direction = op.direction():  ```
 
according to LLVM coding rules, which we strive to follow, we should not use 'auto' everywhere. Specifically auto should not be used when the type of the RHS expression is not clear from the context. Using the actual type helps the reader of the code,  `auto` should be used when either the RHS type is obvious from the context, or the type of the RHS expression is 'complex' (like when declaring a lambda function).  

This is not specific to this PR, but I suggest we progressively review the code (especially the accelerator code because is new in this project) and, whenever possible, fixup this issue so that, over time, we can cleanup the code. 

In this case the type of `direction` is StringRef, which surprised me because I would expect it to be an `enum class` given that there is only a limited set of values a 'direction' can have.  Would it be possible to change it to an enum ?  Using an enum is cleaner IMO and also has a performance benefit because it is less expensive to do comparisons against enum values that it is to compare strings.

I also suggest to change the `if` into an assert because `direction` can only be one of FORWARD, REVERSE, or BIDIRECTIONAL, so any other value should not be possible. What do you think?

Reply:
I removed "auto" from the declaration. Thank you.

As for use of `enum class`, do you mean that we should use `enum class` in ZHighOps.td and ZLowOps.td,
or we should use `enum class` only in C++ programs?
To use `enum class` in ZHighOps.td and ZLowOps.td, large code modification is required, because the current ZHighOps.td and ZLowOps.td use `StringAttr` for several places (e.g. layout, padding_type, direction and act_func), so that I will create another PR. Can you give me comments or suggestions?

Reply:
> I removed "auto" from the declaration. Thank you.

Thanks @negiyas. 

> As for use of enum class, do you mean that we should use enum class in ZHighOps.td and ZLowOps.td,
or we should use enum class only in C++ programs?
To use enum class in ZHighOps.td and ZLowOps.td, large code modification is required, because the current ZHighOps.td and ZLowOps.td use StringAttr for several places (e.g. layout, padding_type, direction and act_func), so that I will create another PR. Can you give me comments or suggestions?

As I mentioned in this thread I find unintuitive that `direction` is a string, because it has a limited number of possible values, which could be modeled using an enumeration (either a pre C++11 `enum` or, preferably, a C++11 enum class). Using enumerators would allow us to avoid string comparisons which are more costly.

I checked the ONNX specification for LSTM (other operators like GRU are similar). The `direction` attribute is defined to be a string:

```
direction : string (default is forward)
Specify if the RNN is forward, reverse, or bidirectional. Must be one of forward (default), reverse, or bidirectional.
```

So the current code "mirrors" what is in the ONNX specification closely. Internally we could still model the direction as an enum, but I realize this would be a pervasive change. 

In summary, I feel we should review the internal implementation for the direction attribute, but I agree at this point we should do that in a future pull request.